src/cargo/ops/cargo_rustc/mod.rs

```diff
@@ -319,36 +323,30 @@ fn rustc(cx: &mut Context, unit: &Unit) -> CargoResult<Work> {
         // hard link our outputs out of the `deps` directory into the directory
         // above. This means that `cargo build` will produce binaries in
         // `target/debug` which one probably expects.
-        if move_outputs_up {
-            for &(ref filename, _linkable) in filenames.iter() {
-                let src = root.join(filename);
-                // This may have been a `cargo rustc` command which changes the
-                // output, so the source may not actually exist.
-                if !src.exists() {
-                    continue
-                }
+        for (src, link_dst, _linkable) in filenames {
+            // This may have been a `cargo rustc` command which changes the
+            // output, so the source may not actually exist.
+            debug!("Thinking about linking {} to {:?}", src.display(), link_dst);
+            if !src.exists() || link_dst.is_none() {
+                continue
+            }
+            let dst = link_dst.unwrap();
 
-                // We currently only lift files up from the `deps` directory. If
-                // it was compiled into something like `example/` or `doc/` then
-                // we don't want to link it up.
-                let src_dir = src.parent().unwrap();
-                if !src_dir.ends_with("deps") {
-                    continue
-                }
-                let dst = src_dir.parent().unwrap()
-                                 .join(src.file_name().unwrap());
-                if dst.exists() {
-                    try!(fs::remove_file(&dst).chain_error(|| {
-                        human(format!("failed to remove: {}", dst.display()))
-                    }));
-                }
-                try!(fs::hard_link(&src, &dst)
-                     .or_else(|_| fs::copy(&src, &dst).map(|_| ()))
-                     .chain_error(|| {
-                         human(format!("failed to link or copy `{}` to `{}`",
-                                       src.display(), dst.display()))
+            debug!("linking {} to {}", src.display(), dst.display());
+            if dst.exists() {
+                try!(fs::remove_file(&dst).chain_error(|| {
+                    human(format!("failed to remove: {}", dst.display()))
                 }));
             }
+            try!(fs::hard_link(&src, &dst)
```

Comment:
This works today because the fingerprint filename is the same for both deps. Thus if you change the features and try to recompile, it will trigger as dirty. I changed it to have different fingerprint filename for each of these variants so when switching back and forth we keep it working right.